packages/mongodb-memory-server-core/src/util/MongoInstance.ts

```diff
@@ -87,90 +104,97 @@ export default class MongoInstance {
+    if (!!this.instanceOpts.storageEngine) {
+      result.push('--storageEngine', this.instanceOpts.storageEngine);
     }
-    if (dbPath) {
-      result.push('--dbpath', dbPath);
+    if (!!this.instanceOpts.ip) {
+      result.push('--bind_ip', this.instanceOpts.ip);
     }
-    if (!auth) {
-      result.push('--noauth');
-    } else if (auth) {
+    if (this.instanceOpts.auth) {
       result.push('--auth');
+    } else {
+      result.push('--noauth');
     }
-    if (replSet) {
-      result.push('--replSet', replSet);
+    if (!!this.instanceOpts.replSet) {
+      result.push('--replSet', this.instanceOpts.replSet);
     }
 
-    return result.concat(args ?? []);
+    return result.concat(this.instanceOpts.args ?? []);
   }
 
   /**
    * Create the mongod process
    */
   async run(): Promise<this> {
-    const launch = new Promise((resolve, reject) => {
-      this.instanceReady = () => {
-        this.isInstanceReady = true;
-        this.debug('MongodbInstance: Instance is ready!');
-        resolve({ ...this.childProcess });
-      };
-      this.instanceFailed = (err: any) => {
-        this.debug(`MongodbInstance: Instance has failed: ${err.toString()}`);
-        if (this.killerProcess) {
-          this.killerProcess.kill();
-        }
-        reject(err);
-      };
+    const launch: Promise<void> = new Promise((resolve, reject) => {
+      this.once(MongoInstanceEvents.instanceReady, resolve);
+      this.once(MongoInstanceEvents.instanceError, reject);
+      this.once(MongoInstanceEvents.instanceClosed, () => {
+        reject(new Error('Instance Exited before being ready and without throwing an error!'));
+      });
     });
 
-    const mongoBin = await MongoBinary.getPath(this.opts.binary);
+    const mongoBin = await MongoBinary.getPath(this.binaryOpts);
     this.childProcess = this._launchMongod(mongoBin);
     this.killerProcess = this._launchKiller(process.pid, this.childProcess.pid);
 
     await launch;
+    this.emit(MongoInstanceEvents.instanceStarted);
     return this;
   }
 
+  /**
+   * Shutdown all related processes (Mongod Instance & Killer Process)
+   */
   async kill(): Promise<MongoInstance> {
     this.debug('Called MongoInstance.kill():');
 
     /**
      * Function to De-Duplicate Code
-     * @param process The Process to kill
+     * @param childprocess The Process to kill
      * @param name the name used in the logs
      * @param debug the debug function
      */
-    async function kill_internal(process: ChildProcess, name: string, debug: DebugFn) {
-      await new Promise((resolve) => {
-        const timeout = setTimeout(() => {
-          debug('kill_internal timeout triggered');
-          resolve();
-        }, 1000 * 10);
-        process.once(`exit`, (code, signal) => {
-          debug(`- ${name}: got exit signal, Code: ${code}, Signal: ${signal}`);
+    async function kill_internal(this: MongoInstance, childprocess: ChildProcess, name: string) {
```

Comment:
Better to move out this function to utils.
This helps to 
- remove `this` and `call` call
- add more readability (removing function in function)

Reply:
should i also change the `this.debug` to have an parameter (and log as `MongoMS:MongoInstance`) or should it log as `MongoMS:db_utils`?

Reply:
for now i made it to `MongoMS:db_utils` and renamed the function to `killProcess` (instead of `kill_internal`)

Reply:
I think it will be better to use `MongoMS:utils`. But anyway I don't have strong opinion in this thing.

Reply:
ok, it should always match the file name, i will rename it to `utils` after the PR(s)